AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/CreateKeysHV2.dfy

```diff
@@ -320,141 +171,76 @@ module {:options "/functionSyntax:4" } CreateKeysHV2 {
-    kmsClient: KMS.IKMSClient,
-    kmsConfiguration: KeyStoreTypes.KMSConfiguration,
-    grantTokens: KMS.GrantTokenList,
-    encryptionContext:  map<string, string>,
-    branchKeyContext: map<string, string>
-  )
-    reads kmsClient.History
-    requires KMSKeystoreOperations.HasKeyId(kmsConfiguration) && KmsArn.ValidKmsArn?(KMSKeystoreOperations.GetKeyId(kmsConfiguration))
-    requires old(kmsClient.History.Encrypt) < kmsClient.History.Encrypt
-    requires kmsEncryptRequestForActiveBKI in kmsClient.History.Encrypt[|old(kmsClient.History.Encrypt)|..]
-  {
-    // Verify the KMS request to create the ACTIVE ciphertext was constructed correctly
-    && var activeInput := kmsEncryptRequestForActiveBKI.input;
-    && KMSKeystoreOperations.Compatible?(kmsConfiguration, activeInput.KeyId)
-    && |activeInput.Plaintext| == (Structure.BKC_DIGEST_LENGTH + Structure.AES_256_LENGTH) as int
-    && activeInput.EncryptionContext == Some(encryptionContext)
-    && activeInput.GrantTokens == Some(grantTokens)
-    && kmsEncryptRequestForActiveBKI.output.Success?
-    && kmsEncryptRequestForActiveBKI.output.value.CiphertextBlob.Some?
-    && kmsEncryptRequestForActiveBKI.output.value.KeyId.Some?
-    && kmsEncryptRequestForActiveBKI.output.value.KeyId.value == KMSKeystoreOperations.GetKeyId(kmsConfiguration)
-  }
+    returns (output: Result<KeyStoreTypes.EncryptedHierarchicalKey, Types.Error>)
 
+    requires CryptoAndKMS.ValidState()
+    modifies CryptoAndKMS.Modifies
+    ensures CryptoAndKMS.ValidState()
 
-  twostate predicate WrappedBeaconKeyCreationHV2?(
-    new beaconHistory: KMS.DafnyCallEvent<KMS.EncryptRequest, Result<KMS.EncryptResponse, KMS.Error>>,
-    kmsClient: KMS.IKMSClient,
-    kmsConfiguration: KeyStoreTypes.KMSConfiguration,
-    grantTokens: KMS.GrantTokenList,
-    customEncryptionContext:  map<string, string>,
-    decryptOnlyBranchKeyContext: map<string, string>
-  )
-    reads kmsClient.History
-    requires KMSKeystoreOperations.HasKeyId(kmsConfiguration) && KmsArn.ValidKmsArn?(KMSKeystoreOperations.GetKeyId(kmsConfiguration))
-    requires Structure.BranchKeyContext?(decryptOnlyBranchKeyContext)
-    requires old(kmsClient.History.Encrypt) < kmsClient.History.Encrypt
-    requires beaconHistory in kmsClient.History.Encrypt[|old(kmsClient.History.Encrypt)|..]
+    requires
+      && Structure.BranchKeyContext?(branchKeyContext)
+      && |material| == Structure.AES_256_LENGTH as int
+      && KMSKeystoreOperations.AttemptKmsOperation?(CryptoAndKMS.kmsConfig, branchKeyContext[Structure.KMS_FIELD])
+      && KMSKeystoreOperations.GetKeyId(CryptoAndKMS.kmsConfig) == branchKeyContext[Structure.KMS_FIELD]
+
+    // TODO-HV-2-GA: Update Specification for HV-2 Branch Key Creation
+    ensures output.Success? ==>
+              && |CryptoAndKMS.kms.History.Encrypt| == |old(CryptoAndKMS.kms.History.Encrypt)| + 1
+              && var kmsEvent :=  CryptoAndKMS.kms.History.Encrypt[-1 % |CryptoAndKMS.kms.History.Encrypt|];
```

Comment:
Why not
`Seq.Last(CryptoAndKms.kms.History.Encrypt)`?

Reply:
To me, negative indexing is a more natural and flexible implementation than a sequence of drop lasts.
Even for the last member,  I am inclined to do it.
However, I will happily take a TODO to make a helper method that does negative indexing.